packages/ifpack2/src/Ifpack2_Details_Fic_def.hpp

```diff
@@ -0,0 +1,287 @@
+  int level = params.get<int>("level", 0);
+  double omega = params.get<double>("damping factor", 0.5);
+  double shift = params.get<double>("shift", 0);
+  bool guessFlag = params.get<bool>("guess", true);
+  int blkSz = params.get<int>("block size", 5);
+  //get copy of values array from matrix
+  Kokkos::Impl::Timer copyTimer;
+  CrsArrayReader<Scalar, LocalOrdinal, GlobalOrdinal, Node>::getValues(mat.get(), localValues);
+  crsCopyTime += copyTimer.seconds(); //add to the time spent getting rowptrs/colinds
+  auto nRows = mat->getNodeNumRows();
+  //create a new preconditioner with these params (if already exists, this replaces it)
+  localPrec = Teuchos::rcp(new LocalFIC(localRowPtrs, localColInds, localValues, nRows, nFact, nTrisol, level, omega, shift, guessFlag ? 1 : 0, blkSz));
+  localPrec->initialize();
+  localPrec->compute();
+  computedFlag = true;
+  nComputed++;
+  initTime = localPrec->getInitializeTime();
+  computeTime = localPrec->getComputeTime();
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+bool Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+isComputed() const
+{
+  return computedFlag;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+Teuchos::RCP<const Tpetra::RowMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> >
+Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getMatrix() const
+{
+  return mat;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+int Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getNumInitialize() const
+{
+  return nInit;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+int Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getNumCompute() const
+{
+  return nComputed;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+int Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getNumApply() const
+{
+  return nApply;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+double Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getInitializeTime() const 
+{
+  return initTime;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+double Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getComputeTime() const
+{
+  return computeTime;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+double Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getApplyTime() const
+{
+  return applyTime;
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+void Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+checkLocalIC()
+{
+  localPrec->checkIC();
+}
+
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+void Fic<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+printStatus()
+{
```

Comment:
I would prefer that FastILU have a one-line description() method.  This makes it easy to summarize the solve when it's a smoother in a multigrid method, for example.